pyomo/contrib/solver/tests/solvers/test_ipopt.py

```diff
@@ -28,6 +34,248 @@
+            'is_persistent',
+            'solve',
+            'version',
+            'name',
+        ]
+        method_list = [method for method in dir(opt) if method.startswith('_') is False]
+        self.assertEqual(sorted(expected_list), sorted(method_list))
+
+    def test_default_instantiation(self):
+        opt = ipopt.Ipopt()
+        self.assertFalse(opt.is_persistent())
+        self.assertIsNotNone(opt.version())
+        self.assertEqual(opt.name, 'ipopt')
+        self.assertEqual(opt.CONFIG, opt.config)
+        self.assertTrue(opt.available())
+
+    def test_context_manager(self):
+        with ipopt.Ipopt() as opt:
+            self.assertFalse(opt.is_persistent())
+            self.assertIsNotNone(opt.version())
+            self.assertEqual(opt.name, 'ipopt')
+            self.assertEqual(opt.CONFIG, opt.config)
+            self.assertTrue(opt.available())
+
+    def test_available_cache(self):
+        opt = ipopt.Ipopt()
+        opt.available()
+        self.assertTrue(opt._available_cache[1])
+        self.assertIsNotNone(opt._available_cache[0])
+        # Now we will try with a custom config that has a fake path
+        config = ipopt.IpoptConfig()
+        config.executable = Executable('/a/bogus/path')
+        opt.available(config=config)
+        self.assertFalse(opt._available_cache[1])
+        self.assertIsNone(opt._available_cache[0])
+
+    def test_version_cache(self):
+        opt = ipopt.Ipopt()
+        opt.version()
+        self.assertIsNotNone(opt._version_cache[0])
+        self.assertIsNotNone(opt._version_cache[1])
+        # Now we will try with a custom config that has a fake path
+        config = ipopt.IpoptConfig()
+        config.executable = Executable('/a/bogus/path')
+        opt.version(config=config)
+        self.assertIsNone(opt._version_cache[0])
+        self.assertIsNone(opt._version_cache[1])
+
+    def test_write_options_file(self):
+        # If we have no options, we should get false back
+        opt = ipopt.Ipopt()
+        result = opt._write_options_file('fakename', None)
+        self.assertFalse(result)
+        # Pass it some options that ARE on the command line
+        opt = ipopt.Ipopt(solver_options={'max_iter': 4})
+        result = opt._write_options_file('myfile', opt.config.solver_options)
+        self.assertFalse(result)
+        self.assertFalse(os.path.isfile('myfile.opt'))
+        # Now we are going to actually pass it some options that are NOT on
+        # the command line
+        opt = ipopt.Ipopt(solver_options={'custom_option': 4})
+        with TempfileManager.new_context() as temp:
+            dname = temp.mkdtemp()
+            if not os.path.exists(dname):
+                os.mkdir(dname)
+            filename = os.path.join(dname, 'myfile')
+            result = opt._write_options_file(filename, opt.config.solver_options)
+            self.assertTrue(result)
+            self.assertTrue(os.path.isfile(filename + '.opt'))
```

Comment:
This test made me realize that we don't have a good way to make sure there are not typos in options except to look at the ipopt output. Anyone have good ideas on how to do that? The only thing I can think of is to parse the log file.

Reply:
All good points.